### PR TITLE
Support readOnly element.

### LINKED ISSUE #763: Add readOnly itemControl support
adds readOnly to the itemControl (e.g. textbox etc) - see [here](https://www.hl7.org/fhir/questionnaire-definitions.html)

_This is in effect the same as applying the readonly attribute as supported by HTML5 (not to be confused with disabled)_

**This issue is duplicated by https://github.com/google/android-fhir/issues/817**

### PR DESCRIPTION
**Pull request by santosh-pingle**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #763  

**Description**
If questionnaire item has readOnly element true, then make relevant view/widget disable.

**Alternative(s) considered**
No

**Type**
Choose one: Bug fix 

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read [How to Contribute](https://github.com/google/android-fhir/blob/master/docs/contributing.md)
- [x] I have read the [Developer's guide](https://github.com/google/android-fhir/wiki/Developer's-Guide)
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

### CODE REVIEW

**Comment by santosh-pingle** [comment]
> In the RecyclerView, a read-only viewholder may be later reused for a read/write view. In that case `setViewReadOnly` should be called every time to set appropriate `view.isEnabled` value based on q`uestionnaireItemViewItem.questionnaireItem.readOnly` flag.

Done.

**Comment by santosh-pingle** [comment]
Resolving conflicts.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemViewHolderFactory.kt [hunk 1]
```diff
@@ -108,4 +109,14 @@ interface QuestionnaireItemViewHolderDelegate {
       context
     )
   }
+
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemViewHolderFactory.kt:112]: add a function that should be implemented by each widget:
```
fun setReadOnly(isReadOnly: Boolean)
```

**Review by jingtang10 — CHANGES_REQUESTED** [review]
In the `bind` function in `QuestionnaireItemViewHolder`, you call `setReadOnly` which is defined in `QuestionnaireItemViewHolderDelegate`.

**Comment by santosh-pingle** [comment]
readOnly  is true
Screen Shot 2021-11-15 at 12 19 39 PM

gneder is read only
is active read only
Screen Shot 2021-11-15 at 12 19 07 PM

Q4 is read only
Screen Shot 2021-11-15 at 12 25 50 PM

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemAutoCompleteViewHolderFactory.kt [hunk 2]
```diff
@@ -175,6 +175,16 @@ internal object QuestionnaireItemAutoCompleteViewHolderFactory :
           else validationResult.getSingleStringValidationMessage()
       }
 
+      override fun setViewReadOnly(isReadOnly: Boolean) {
+        for (i in 0 until chipContainer.flexItemCount) {
+          var view = chipContainer.getFlexItemAt(i)
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemAutoCompleteViewHolderFactory.kt:180]:
```suggestion
          val view = chipContainer.getFlexItemAt(i)
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemRadioGroupViewHolderFactory.kt [hunk 3]
```diff
@@ -100,5 +99,12 @@ internal object QuestionnaireItemRadioGroupViewHolderFactory :
           if (validationResult.getSingleStringValidationMessage() == "") null
           else validationResult.getSingleStringValidationMessage()
       }
+
+      override fun setViewReadOnly(isReadOnly: Boolean) {
+        for (i in 0 until radioGroup.childCount) {
+          var view = radioGroup.getChildAt(i)
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireItemRadioGroupViewHolderFactory.kt:105]:
```suggestion
          val view = radioGroup.getChildAt(i)
```

**Review by jingtang10 — APPROVED** [review]
Great work thanks!

**Comment by santosh-pingle** [comment]
@jingtang10 
It looks like we need to handle below use case as well.
-> If questionnaire item has `readOnly` property set to true and it also has nested items, then those all nested item views need to be readonly. And it is required to set `readOnly` property of all those nested item in `QuestionnaireViewModel` if parent item is read only. 
correct?

**Comment by jingtang10** [comment]
> If questionnaire item has `readOnly` property set to true and it also has nested items, then those all nested item views need to be set readonly. To make this change, it is required to set `readOnly` property for all those nested item in `QuestionnaireViewModel`

Where is this from? I couldn't find it either on the questionnaire resource documentation or the sdc ig.

Personally I'd like to understand why this is the case -- it's not inconceivable that you have a top-level answer that cannot be changed with nested answer that can be changed.

I'm happy for you to add a TODO (and create an issue) with a link to the source of this.

**Comment by santosh-pingle** [comment]
> > If questionnaire item has `readOnly` property set to true and it also has nested items, then those all nested item views need to be set readonly. To make this change, it is required to set `readOnly` property for all those nested item in `QuestionnaireViewModel`
> 
> Where is this from? I couldn't find it either on the questionnaire resource documentation or the sdc ig.
> 
> Personally I'd like to understand why this is the case -- it's not inconceivable that you have a top-level answer that cannot be changed with nested answer that can be changed.
> 
> I'm happy for you to add a TODO (and create an issue) with a link to the source of this.

Just for reference,
Discussed offline.
Its not required to handle, there is no such reference, just query which I had.

**Pull request merged**